exporter/graph/graph_crawler.py

```diff
@@ -0,0 +1,55 @@
+from exporter.metadata import MetadataResource, MetadataService
+from exporter.graph.experiment_graph import ExperimentGraph, Link, Input, Output, ProtocolLink
+from typing import List, Iterable, Dict
+from functools import reduce
+from _operator import iconcat
+from dataclasses import dataclass
+
+
+@dataclass
+class ProcessInfo:
+    process: MetadataResource
+    inputs: List[MetadataResource]
+    outputs: List[MetadataResource]
+    protocols: List[MetadataResource]
+
+
+class GraphCrawler:
+    def __init__(self, metadata_service: MetadataService):
+        self.metadata_service = metadata_service
+
+    def experiment_graph_for_process(self, process: MetadataResource) -> ExperimentGraph:
+        return self._crawl(process)
+
+    def _crawl(self, process: MetadataResource) -> ExperimentGraph:
+        partial_graph = ExperimentGraph()
+
+        process_info = self.process_info(process)
+        partial_graph.nodes.add_nodes(process_info.inputs + process_info.outputs + process_info.protocols + [process])
+        partial_graph.links.add_link(GraphCrawler.link_for(process_info))
+
+        processes_to_crawl = GraphCrawler.flatten([self.metadata_service.get_derived_by_processes(i) for i in process_info.inputs])
+
+        return reduce(lambda g1, g2: g1.extend(g2),
+                      map(lambda proc: self._crawl(proc), processes_to_crawl),
+                      partial_graph)
```

Comment:
very concise

Reply:
Cheers. We might be able to use this for driving the archival/manifest-generator process later on as well